pkg/controller/pkg/revision/linter.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"github.com/crossplane/crossplane-runtime/pkg/parser"
+
+	apiextensionsv1alpha1 "github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	pkgmeta "github.com/crossplane/crossplane/apis/pkg/meta/v1alpha1"
+)
+
+const (
+	errMultipleMeta         = "not exactly one package meta type"
+	errNotMetaProvider      = "package meta type is not Provider"
+	errNotMetaConfiguration = "package meta type is not Configuration"
+	errNotCRD               = "object is not a CRD"
+	errNotXRD               = "object is not an XRD"
+	errNotComposition       = "object is not a Composition"
+)
+
+// A Linter lints packages.
```

Comment:
I anticipate moving much of this (excluding the functions specific to Crossplane types) to `crossplane-runtime` at some point